src/app/designs/components/design-modules/design-module/design-module.component.ts

```diff
@@ -0,0 +1,192 @@
+      this.designModule.designId,
+      this.designModule.id
+    );
+  }
+
+  ngOnChanges() {
+    if (
+      this.module &&
+      this.module.versions &&
+      this.designModule.moduleVersion
+    ) {
+      const version = this.module.versions.find(
+        (x) => x.name == this.designModule.moduleVersion
+      );
+
+      if (version) {
+        this.outputsSubject.next(version.outputs);
+      }
+    }
+  }
+
+  delete() {
+    this.confirmationService
+      .confirmDialog(
+        'Delete Module',
+        `Are you sure you want to delete ${this.designModule.name}?`
+      )
+      .pipe(take(1))
+      .subscribe((result) => {
+        if (!result[this.confirmationService.WAS_CANCELLED]) {
+          this.designModuleService.delete(this.designModule.id).subscribe();
+        }
+      });
+  }
+
+  edit() {
+    this.showOutputs = false;
+    this.isEditing = !this.isEditing;
+
+    if (this.isEditing) {
+      // wait on first load, otherwise refresh in background
+      // in case signalR missed an update
+      if (!this.designModule.values || !this.module.versions) {
+        this.loading = true;
+      }
+
+      forkJoin({
+        designModule: this.designModuleService.load(this.designModule.id),
+        module: this.moduleService.loadModuleById(this.module.id),
+      }).subscribe(() => {
+        this.loading = false;
+      });
+    }
+  }
+
+  toggleOutputs() {
+    this.isEditing = false;
+    this.showOutputs = !this.showOutputs;
+
+    if (this.showOutputs) {
+      this.loading = true;
+      this.moduleService.loadModuleById(this.module.id).subscribe(() => {
+        this.loading = false;
+      });
+    }
+  }
+
+  toggleEnabled() {
+    this.designModuleService.toggleEnabled(this.designModule);
+  }
+
+  onEditComplete(event: ModuleVariablesResult) {
+    if (!event) {
+      this.isEditing = false;
+      return;
+    }
+
+    let apiCall: Observable<DesignModule>;
+
+    const changedValues = event.variableValues.filter((x) =>
+      event.changedVariables.includes(x.name)
+    );
+
+    // edit the designModule if any of it's properties changed
+    if (
+      event.moduleName != this.designModule.name ||
+      event.versionName != this.designModule.moduleVersion
+    ) {
+      const command = {
+        designId: this.designModule.designId,
+        moduleId: this.module.id,
+        moduleVersion: event.versionName,
+        name: event.moduleName,
+        values: changedValues,
+      } as EditDesignModuleCommand;
+
+      apiCall = this.designModuleService.edit(this.designModule.id, command);
+    } else if (event.changedVariables.length > 0) {
+      // only update the designModule's variable values
+      apiCall = this.designModuleService.addOrUpdateValues(
+        this.designModule.id,
+        {
+          values: changedValues,
+        }
+      );
+    }
+
+    if (apiCall) {
+      this.saving = true;
+      apiCall.subscribe(() => (this.saving = false));
```

Comment:
Missing takeUntil(this.unsubscribe$)

Reply:
It's not necessary to unsubscribe from an httpClient observable, since it can only ever emit a single value and then calls complete.

refs:
https://lukaonik.medium.com/do-we-need-to-unsubscribe-http-client-in-angular-86d781522b99